File: C++/9-DynamicProgramming(DP)/Problems/4-LongestPalindromicSubstring/LongestPalindromicSubstringPro.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <limits> // Required for numeric_limits

using namespace std;

// Function to print the Longest Palindromic Substring and its length
void printResult(const string& s, int start, int end) {
    cout << "Longest Palindromic Substring: ";
    for (int i = start; i <= end; ++i) {
        cout << s[i];
    }
    cout << endl;
    cout << "Length: " << end - start + 1 << endl;
}

// 1. Brute Force Approach
//   - Iterate through all possible substrings and check if each is a palindrome.
//   - Time Complexity: O(n^3), Space Complexity: O(1)
string longestPalindromeBruteForce(const string& s) {
    int n = s.length();
    if (n == 0) return "";
    int start = 0, maxLength = 1;

    for (int i = 0; i < n; ++i) {
        for (int j = i; j < n; ++j) {
            bool isPalindrome = true;
            for (int k = 0; k < (j - i + 1) / 2; ++k) {
                if (s[i + k] != s[j - k]) {
                    isPalindrome = false;
                    break;
                }
            }
            if (isPalindrome && (j - i + 1) > maxLength) {
                start = i;
                maxLength = j - i + 1;
            }
        }
    }
    return s.substr(start, maxLength);
}

// 2. Dynamic Programming Approach
//   - Create a 2D table where dp[i][j] is true if the substring s[i...j] is a palindrome.
//   - Build the table bottom-up.
//   - Time Complexity: O(n^2), Space Complexity: O(n^2)
string longestPalindromeDP(const string& s) {
    int n = s.length();
    if (n == 0) return "";
    vector<vector<bool>> dp(n, vector<bool>(n, false));
    int start = 0, maxLength = 1;

    // Base cases: single characters are palindromes
    for (int i = 0; i < n; ++i) {
        dp[i][i] = true;
    }

    // Check for substrings of length 2
    for (int i = 0; i < n - 1; ++i) {
        if (s[i] == s[i + 1]) {
            dp[i][i + 1] = true;
            start = i;
            maxLength = 2;
        }
    }

    // Check for substrings of length 3 or greater
    for (int k = 3; k <= n; ++k) {
        for (int i = 0; i < n - k + 1; ++i) {
            int j = i + k - 1; // Ending index of the substring
            if (s[i] == s[j] && dp[i + 1][j - 1]) {
                dp[i][j] = true;
                if (k > maxLength) {
                    start = i;
                    maxLength = k;
                }
            }
        }
    }
    return s.substr(start, maxLength);
}

// 3. Expand Around Center Approach
//   - For each character in the string, consider it as the center of a palindrome and expand in both directions.
//   - Handle both odd and even length palindromes.
//   - Time Complexity: O(n^2), Space Complexity: O(1)
string longestPalindromeExpandAroundCenter(const string& s) {
    int n = s.length();
    if (n == 0) return "";
    int start = 0, end = 0;

    for (int i = 0; i < n; ++i) {
        // For odd length palindromes
        int left = i, right = i;
        while (left >= 0 && right < n && s[left] == s[right]) {
            if (right - left + 1 > end - start + 1) {
                start = left;
                end = right;
            }
            left--;
            right++;
        }

        // For even length palindromes
        left = i, right = i + 1;
        while (left >= 0 && right < n && s[left] == s[right]) {
            if (right - left + 1 > end - start + 1) {
                start = left;
                end = right;
            }
            left--;
            right++;
        }
    }
    return s.substr(start, end - start + 1);
}

// 4. Manacher's Algorithm
//   - A linear time algorithm to find the longest palindromic substring.
//   - It uses a transformed string and an array to store palindrome lengths.
//   - Time Complexity: O(n), Space Complexity: O(n)
string longestPalindromeManacher(const string& s) {
    int n = s.length();
    if (n == 0) return "";

    // Transform the string to handle even and odd length palindromes uniformly
    string t = "^"; // Start boundary
    for (int i = 0; i < n; ++i) {
        t += "#";
        t += s[i];
    }
    t += "#$"; // End boundary
    int m = t.length();

    vector<int> p(m, 0); // Array to store palindrome lengths at each center
    int center = 0, right = 0;
    int start = 0, maxLength = 0;

    for (int i = 1; i < m - 1; ++i) {
        int mirror = 2 * center - i; // Mirror position of i with respect to center

        if (i < right) {
            p[i] = min(right - i, p[mirror]); // Take minimum of mirror's value and right boundary
        }

        // Expand around center i
        while (t[i + (1 + p[i])] == t[i - (1 + p[i])]) {
            p[i]++;
        }

        // If current palindrome expands past right, adjust center and right
        if (i + p[i] > right) {
            center = i;
            right = i + p[i];
        }

        // Update maxLength and start index
        if (p[i] > maxLength) {
            maxLength = p[i];
            start = (i - maxLength) / 2; // Original start position
        }
    }
    return s.substr(start, maxLength);
}

// 5. Optimized Dynamic Programming Approach (Space Optimized)
//   - Similar to the DP approach, but optimizes space complexity to O(n) by using only two rows of the DP table.
//   - Time Complexity: O(n^2), Space Complexity: O(n)
string longestPalindromeOptimizedDP(const string& s) {
    int n = s.length();
    if (n == 0) return "";
    vector<bool> prevRow(n, false);
    vector<bool> currRow(n, false);
    int start = 0, maxLength = 1;

    // Base case: single characters are palindromes
    for (int i = 0; i < n; ++i) {
        prevRow[i] = true;
    }

    // Check for substrings of length 2
    for (int i = 0; i < n - 1; ++i) {
        if (s[i] == s[i + 1]) {
            currRow[i] = true;
            start = i;
            maxLength = 2;
        } else {
            currRow[i] = false;
        }
    }
    prevRow = currRow; // Copy currRow to prevRow for next iteration

    // Check for substrings of length 3 or greater
    for (int k = 3; k <= n; ++k) {
        currRow.assign(n, false); // Reset currRow for each k
        for (int i = 0; i < n - k + 1; ++i) {
            int j = i + k - 1;
            if (s[i] == s[j] && prevRow[i + 1]) {
                currRow[i] = true;
                if (k > maxLength) {
                    start = i;
                    maxLength = k;
                }
            }
        }
        prevRow = currRow; // Copy currRow to prevRow for next iteration
    }
    return s.substr(start, maxLength);
}

int main() {
    string s;
    cout << "Enter a string: ";
    cin >> s;

    cout << "\n--- Brute Force Approach ---" << endl;
    string resultBruteForce = longestPalindromeBruteForce(s);
    printResult(s, s.find(resultBruteForce), s.find(resultBruteForce) + resultBruteForce.length() - 1);

    cout << "\n--- Dynamic Programming Approach ---" << endl;
    string resultDP = longestPalindromeDP(s);
    printResult(s, s.find(resultDP), s.find(resultDP) + resultDP.length() - 1);

    cout << "\n--- Expand Around Center Approach ---" << endl;
    string resultExpandAroundCenter = longestPalindromeExpandAroundCenter(s);
    printResult(s, s.find(resultExpandAroundCenter), s.find(resultExpandAroundCenter) + resultExpandAroundCenter.length() - 1);

    cout << "\n--- Manacher's Algorithm ---" << endl;
    string resultManacher = longestPalindromeManacher(s);
    printResult(s, s.find(resultManacher), s.find(resultManacher) + resultManacher.length() - 1);

    cout << "\n--- Optimized Dynamic Programming Approach ---" << endl;
    string resultOptimizedDP = longestPalindromeOptimizedDP(s);
    printResult(s, s.find(resultOptimizedDP), s.find(resultOptimizedDP) + resultOptimizedDP.length() - 1);

    return 0;
}

File: Java/9-DynamicProgramming(DP)/Problems/4-LongestPalindromicSubstring/LongestPalindromicSubstringPro.java
import java.util.Arrays;

public class LongestPalindromicSubstring {

    /**
     * Approach 1: Brute Force
     * -   Checks every possible substring for being a palindrome.
     * -   Time Complexity: O(n^3), where n is the length of the string.
     * -   Space Complexity: O(1).
     * -   Real-world application: Simple string processing in low-resource environments
     * where code simplicity is prioritized over performance, and input sizes are small.
     * For example, checking user input for palindromic properties in a very basic application.
     */
    public static String longestPalindromeBruteForce(String s) {
        if (s == null || s.length() < 1) return "";
        String longest = "";
        for (int i = 0; i < s.length(); i++) {
            for (int j = i; j < s.length(); j++) {
                String sub = s.substring(i, j + 1);
                if (isPalindrome(sub) && sub.length() > longest.length()) {
                    longest = sub;
                }
            }
        }
        return longest;
    }

    private static boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    /**
     * Approach 2: Dynamic Programming
     * -   Creates a 2D table where dp[i][j] is true if the substring s[i...j] is a palindrome.
     * -   Time Complexity: O(n^2).
     * -   Space Complexity: O(n^2).
     * -   Real-world application: Bioinformatics for sequence alignment, text editing
     * software for advanced text analysis, and data compression algorithms.
     */
    public static String longestPalindromeDP(String s) {
        if (s == null || s.length() < 1) return "";
        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        int start = 0;
        int maxLen = 1;

        // Base cases: single characters are palindromes
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
        }

        // Check for palindromes of length 2
        for (int i = 0; i < n - 1; i++) {
            if (s.charAt(i) == s.charAt(i + 1)) {
                dp[i][i + 1] = true;
                start = i;
                maxLen = 2;
            }
        }

        // Check for palindromes of length greater than 2
        for (int len = 3; len <= n; len++) {
            for (int i = 0; i <= n - len; i++) {
                int j = i + len - 1;
                if (s.charAt(i) == s.charAt(j) && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    if (len > maxLen) {
                        start = i;
                        maxLen = len;
                    }
                }
            }
        }
        return s.substring(start, start + maxLen);
    }


    /**
     * Approach 3: Expand Around Center
     * -   For each character, expand outwards to find the longest palindrome with that character as the center.
     * -   Time Complexity: O(n^2).
     * -   Space Complexity: O(1).
     * -    Real-world application: Used in text processing applications, such as
     * detecting palindromic words or phrases in a document, and in bioinformatics
     * for analyzing DNA sequences.
     */
    public static String longestPalindromeExpandAroundCenter(String s) {
        if (s == null || s.length() < 1) return "";
        int start = 0;
        int maxLen = 1;
        for (int i = 0; i < s.length(); i++) {
            // For odd length palindromes
            int len1 = expandAroundCenter(s, i, i);
            // For even length palindromes
            int len2 = expandAroundCenter(s, i, i + 1);
            int len = Math.max(len1, len2);
            if (len > maxLen) {
                maxLen = len;
                start = i - (len - 1) / 2;
            }
        }
        return s.substring(start, start + maxLen);
    }

    private static int expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }

    /**
     * Approach 4: Manacher's Algorithm
     * -   A linear time complexity algorithm to find the longest palindromic substring.
     * -   Time Complexity: O(n).
     * -   Space Complexity: O(n).
     * -   Real-world application: High-performance text processing, advanced bioinformatics
     * research, and specialized string manipulation libraries.  This is the most
     * efficient algorithm, but it's also the most complex.
     */
    public static String longestPalindromeManacher(String s) {
        if (s == null || s.length() < 1) return "";
        // Preprocess the string to handle even length palindromes
        String processedS = preprocess(s);
        int n = processedS.length();
        int[] p = new int[n]; // Array to store palindrome lengths at each center
        int center = 0, right = 0; // center and right boundary of the rightmost palindrome
        int start = 0, maxLen = 1;

        for (int i = 1; i < n - 1; i++) {
            int mirror = 2 * center - i; // Mirror position of i with respect to center
            if (i < right) {
                p[i] = Math.min(right - i, p[mirror]); // Take minimum of mirror's length and right boundary
            }

            // Expand around center i
            while (processedS.charAt(i + (1 + p[i])) == processedS.charAt(i - (1 + p[i]))) {
                p[i]++;
            }

            // If current palindrome expands past right, update center and right
            if (i + p[i] > right) {
                center = i;
                right = i + p[i];
            }

            // Update maxLen and start if current palindrome is longer
            if (p[i] > maxLen) {
                maxLen = p[i];
                start = (i - maxLen) / 2; // Original start index in s
            }
        }
        return s.substring(start, start + maxLen);
    }

    private static String preprocess(String s) {
        if (s == null || s.length() == 0) return "^$";  // Handle empty string case
        StringBuilder sb = new StringBuilder("^"); // Start with a dummy character
        for (int i = 0; i < s.length(); i++) {
            sb.append("#").append(s.charAt(i)); // Insert '#' between characters
        }
        sb.append("#$"); // End with '#' and a dummy character
        return sb.toString();
    }

    /**
     * Approach 5: Optimized Dynamic Programming (Space Optimized)
     * -   Optimizes the space complexity of the DP approach by using only two rows
     * of the DP table at a time.
     * -   Time Complexity: O(n^2).
     * -   Space Complexity: O(n).
     * -   Real-world application: Mobile applications or embedded systems with limited memory.
     */
    public static String longestPalindromeOptimizedDP(String s) {
        if (s == null || s.length() < 1) return "";
        int n = s.length();
        boolean[][] dp = new boolean[2][n]; // Only two rows are needed
        int start = 0;
        int maxLen = 1;
        int currentRow = 0;

        // Base cases: single characters are palindromes
        for (int i = 0; i < n; i++) {
            dp[currentRow][i] = true;
        }

        // Check for palindromes of length 2
        for (int i = 0; i < n - 1; i++) {
            if (s.charAt(i) == s.charAt(i + 1)) {
                dp[currentRow][i] = true;
                start = i;
                maxLen = 2;
            } else {
                dp[currentRow][i] = false; // Important:  Need to set this for the loop below
            }
        }

        // Check for palindromes of length greater than 2
        for (int len = 3; len <= n; len++) {
            int prevRow = currentRow; // Store the previous row
            currentRow = 1 - currentRow; // Switch to the other row (0 or 1)

            for (int i = 0; i <= n - len; i++) {
                int j = i + len - 1;
                if (s.charAt(i) == s.charAt(j) && dp[prevRow][i + 1]) {
                    dp[currentRow][i] = true;
                    if (len > maxLen) {
                        start = i;
                        maxLen = len;
                    }
                } else {
                    dp[currentRow][i] = false;
                }
            }
        }
        return s.substring(start, start + maxLen);
    }

    public static void main(String[] args) {
        String testString = "babad";
        System.out.println("Input string: " + testString);

        System.out.println("\nLongest Palindromic Substring - Brute Force: " + longestPalindromeBruteForce(testString));
        System.out.println("Longest Palindromic Substring - Dynamic Programming: " + longestPalindromeDP(testString));
        System.out.println("Longest Palindromic Substring - Expand Around Center: " + longestPalindromeExpandAroundCenter(testString));
        System.out.println("Longest Palindromic Substring - Manacher's Algorithm: " + longestPalindromeManacher(testString));
        System.out.println("Longest Palindromic Substring - Optimized DP: " + longestPalindromeOptimizedDP(testString));

        String testString2 = "cbbd";
        System.out.println("\nInput string: " + testString2);

        System.out.println("\nLongest Palindromic Substring - Brute Force: " + longestPalindromeBruteForce(testString2));
        System.out.println("Longest Palindromic Substring - Dynamic Programming: " + longestPalindromeDP(testString2));
        System.out.println("Longest Palindromic Substring - Expand Around Center: " + longestPalindromeExpandAroundCenter(testString2));
        System.out.println("Longest Palindromic Substring - Manacher's Algorithm: " + longestPalindromeManacher(testString2));
        System.out.println("Longest Palindromic Substring - Optimized DP: " + longestPalindromeOptimizedDP(testString2));

        String testString3 = "a";
        System.out.println("\nInput string: " + testString3);
        System.out.println("\nLongest Palindromic Substring - Brute Force: " + longestPalindromeBruteForce(testString3));
        System.out.println("Longest Palindromic Substring - Dynamic Programming: " + longestPalindromeDP(testString3));
        System.out.println("Longest Palindromic Substring - Expand Around Center: " + longestPalindromeExpandAroundCenter(testString3));
        System.out.println("Longest Palindromic Substring - Manacher's Algorithm: " + longestPalindromeManacher(testString3));
        System.out.println("Longest Palindromic Substring - Optimized DP: " + longestPalindromeOptimizedDP(testString3));

        String testString4 = "ac";
        System.out.println("\nInput string: " + testString4);
        System.out.println("\nLongest Palindromic Substring - Brute Force: " + longestPalindromeBruteForce(testString4));
        System.out.println("Longest Palindromic Substring - Dynamic Programming: " + longestPalindromeDP(testString4));
        System.out.println("Longest Palindromic Substring - Expand Around Center: " + longestPalindromeExpandAroundCenter(testString4));
        System.out.println("Longest Palindromic Substring - Manacher's Algorithm: " + longestPalindromeManacher(testString4));
        System.out.println("Longest Palindromic Substring - Optimized DP: " + longestPalindromeOptimizedDP(testString4));
    }
}

